[RNMobile] [Embed block] Error bottom sheet

jd-alexander opened the pull request:
`gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3921

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Adds support for backend and network-related error resolution. This resolution involves presenting options to the user that will allow them to recover from the error state. 

#### For network related errors while fetching the preview (timeout, no internet connection)
- Bottom shows error message and **only** the option to retry
- When pressed try to load the preview for the same embed URL again

#### For backend errors while fetching the preview (unsupported embed URL, or cases where a retry wouldn't change the result)
- Bottom shows error message and **only** the option to convert to link
- When pressed Embed block should be converted to a paragraph block that includes the unsupported URL

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

### Unembeddable URL

1. Add an Embed block. 
2. Insert an unembeddable URL like `http://example.com`. 
3. Notice that the Error UI state is shown. 

iOS | Android 
--------|-------
  <kbd><img src="https://user-images.githubusercontent.com/1509205/132262342-aaea7597-1246-4f57-b3d8-24dc73c1826e.png" width="320"></kbd>     |     <kbd><img src="https://user-images.githubusercontent.com/1509205/132262369-9910040c-27c5-42bb-a91b-5a61770ecada.png" width="320"></kbd>
  

### Retry
4. Press **Retry**
5. The Embed block should be reloaded. 

### Convert to link
6. Press **Convert to link**
7. The Embed block should be converted to a **Paragraph** block that has the URL as its contents. 

iOS | Android 
--------|-------
  <kbd><img src="https://user-images.githubusercontent.com/1509205/132262488-d19a9dfb-1ec4-43a0-8b79-143fcf9f689b.gif" width="320"></kbd>     |     <kbd><img src="https://user-images.githubusercontent.com/1509205/132262486-96351633-0642-42a4-81a1-50385e419d1e.gif" width="320"></kbd>
  
  ## Concerns ⚠️ 
1. When the device does not have an internet connection, no **timeout** occurs. The Embed block remains in the **Loading** UI state. 

iOS | Android 
--------|-------
  <kbd><img src="https://user-images.githubusercontent.com/1509205/132262637-2146793c-76b1-48a9-b3c6-4ce7ab44188e.gif" width="320"></kbd>     |     <kbd><img src="https://user-images.githubusercontent.com/1509205/132262636-159f9059-1c3d-4a87-a54d-193048aabc47.gif" width="320"></kbd>

2. The description states that the Retry option should only be available for Embeds that can't be loaded due to a network-related error. This will be possible once the **timeout** issue has been resolved. 

3. The design contains a `Picker` that has a description. The current `Picker` implementation does not support this hence I opted to show only the error description which is a generic hardcoded value that the web Embed component allows displays. Also, I noticed that the existing Error UI in the Placeholder is **Red** instead of **Dark Grey**. Should this be changed? 

Design | Implementation 
--------|-------
<kbd><img src="https://user-images.githubusercontent.com/1509205/132262888-ded29ad6-843f-4cdc-b69b-09ca6d428f88.png" width="320"></kbd> |  <kbd><img src="https://user-images.githubusercontent.com/1509205/132262342-aaea7597-1246-4f57-b3d8-24dc73c1826e.png" width="320"></kbd>     
  


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
* Implemented Embed block error use cases utilizing the mobile `Picker` component. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jd-alexander committed:
Added the initial implementation of the  embed error picker.

jd-alexander committed:
Added a comment to clarify the necessity of the resolveOnPressEvent func

jd-alexander committed:
Integrated the fallback and tryAgain behavior with the EmbedPlaceholder

jd-alexander committed:
Added an entry to CHANGELOG.

jd-alexander committed:
Added PR number to CHANGELOG entry.

jd-alexander commented:
### Connectivity Error

To provide more context here, I investigated the `fetching` issue that would cause the **Loading** UI state. What I have discovered is 

https://github.com/WordPress/gutenberg/blob/dd5fd6b2466d33177d46dcdc0733abd0217e044a/packages/block-library/src/embed/edit.native.js#L87

`embedPreview` is undefined when no internet is available. 

Therefore the logic below will always pass, as `isRequestingEmbedPreview` returns `true` and `getEmbedPreview` is `undefined` so the second part of the condition would do the same. 

https://github.com/WordPress/gutenberg/blob/dd5fd6b2466d33177d46dcdc0733abd0217e044a/packages/block-library/src/embed/edit.native.js#L105

### Server Error 

- `getEmbedPreview` iOS output: 
```
[Error: Error Domain=NSURLErrorDomain Code=404 "The operation couldn’t be completed. (WordPressKit.WordPressComRestApiError error 7.)" UserInfo={NSLocalizedDescription=The operation couldn’t be completed. (WordPressKit.WordPressComRestApiError error 7.)}]
```

- `getEmbedPreview` Android output : 

```
[Error: null]
```

-------

So in the case of the Server Error, there's the `Error` output instead of `undefined` when `getEmbedPreview` is called. Would we need to propagate something like this from the network layer so we can handle the internet connectivity case in a better way?

This would be helpful in crafting a Boolean `serverError` `prop` that could be passed to the `EmbedPlaceholder` to determine if the `Retry` action should be shown. 

Hence, the reason I am calculating which options should be available in the `Picker` 
https://github.com/WordPress/gutenberg/blob/7b31b96760b490dfbda6cb505534671c0d7b58d7/packages/block-library/src/embed/embed-placeholder.native.js#L68

fluiddot commented:
> ### Connectivity Error
> To provide more context here, I investigated the `fetching` issue that would cause the **Loading** UI state. What I have discovered is
> 
> https://github.com/WordPress/gutenberg/blob/dd5fd6b2466d33177d46dcdc0733abd0217e044a/packages/block-library/src/embed/edit.native.js#L87
> 
> `embedPreview` is undefined when no internet is available.
> 
> Therefore the logic below will always pass, as `isRequestingEmbedPreview` returns `true` and `getEmbedPreview` is `undefined` so the second part of the condition would do the same.
> 
> https://github.com/WordPress/gutenberg/blob/dd5fd6b2466d33177d46dcdc0733abd0217e044a/packages/block-library/src/embed/edit.native.js#L105
> 
> ### Server Error
> * `getEmbedPreview` iOS output:
> 
> ```
> [Error: Error Domain=NSURLErrorDomain Code=404 "The operation couldn’t be completed. (WordPressKit.WordPressComRestApiError error 7.)" UserInfo={NSLocalizedDescription=The operation couldn’t be completed. (WordPressKit.WordPressComRestApiError error 7.)}]
> ```
> 
> * `getEmbedPreview` Android output :
> 
> ```
> [Error: null]
> ```
> 
> So in the case of the Server Error, there's the `Error` output instead of `undefined` when `getEmbedPreview` is called. Would we need to propagate something like this from the network layer so we can handle the internet connectivity case in a better way?
> 
> This would be helpful in crafting a Boolean `serverError` `prop` that could be passed to the `EmbedPlaceholder` to determine if the `Retry` action should be shown.
> 
> Hence, the reason I am calculating which options should be available in the `Picker`
> https://github.com/WordPress/gutenberg/blob/7b31b96760b490dfbda6cb505534671c0d7b58d7/packages/block-library/src/embed/embed-placeholder.native.js#L68

Awesome detective work ❤️ !

Looks like we need to review the error handling on the native side because we should receive at least the status code when the request fails. I also noticed that the [`api-fetch-setup.js file`](https://github.com/WordPress/gutenberg/blob/2c56a9cbd2530def2840221e60b113c645acd74c/packages/react-native-editor/src/api-fetch-setup.js#L37), which processes the requests initiated from the RN side, expects the status code for handling the error so this might be a regression 🤔 .

At this point, we should focus on solving the missing error status code and figure out a way to detect the availability of the internet connection to cover the network-related errors. But I don't think this should block the PR as we could address it when tackling the [Offline support for Embeds blocks issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3908), wdyt?

fluiddot commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ For each user feature we should also add a importance categorization label  to i
 -->
 
 ## Unreleased
+-   [**] Embed block : Add error bottom sheet with retry and convert to link actions. [#34604]
```
I noticed that there's an extra space between `block` and `:`.

jd-alexander replied:
Thanks for spotting. Resolved 91e8582

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -40,12 +44,58 @@ const EmbedPlaceholder = ( {
 		<>
 			<TouchableWithoutFeedback
 				accessibilityRole={ 'button' }
 				accessibilityHint={ __( 'Double tap to add a link.' ) }
```
I think we should display a different message here if the preview can't be embedded, wdyt?

jd-alexander replied:
I agree. Implemented here 06a0eb4

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -40,12 +44,58 @@ const EmbedPlaceholder = ( {
+	}
+
+	function onPickerPresent() {
+		if ( cannotEmbedMenuPickerRef.current ) {
+			cannotEmbedMenuPickerRef.current.presentPicker();
+		}
```
We could simplify this block with the following line:
```suggestion
		cannotEmbedMenuPickerRef.current?.presentPicker();
```

I also see that this function (`onPickerPresent`) is only called from `resolveOnPressEvent` so we might consider removing it and move the code there, wdyt?

jd-alexander replied:
Great observation! Done here deb2eb7

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -40,12 +44,58 @@ const EmbedPlaceholder = ( {
+		}
+	}
+
+	// when the content cannot be embedded the onPress should trigger the Picker instead of the onPress prop.
```
We usually capitalize the first letter of a comment so we could change this to:
```suggestion
	// When the content cannot be embedded the onPress should trigger the Picker instead of the onPress prop.
```

jd-alexander replied:
Done here deb2eb7

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -40,12 +44,58 @@ const EmbedPlaceholder = ( {
+			id: 'retryOption',
+			label: __( 'Retry' ),
+			value: 'retryOption',
+			onSelect: () => {
+				tryAgain();
+			},
```
We could simplify this line by doing:
```suggestion
			onSelect: tryAgain,
```

jd-alexander replied:
Agreed! a2dc77b

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -40,12 +44,58 @@ const EmbedPlaceholder = ( {
+			id: 'convertToLinkOption',
+			label: __( 'Convert to link' ),
+			value: 'convertToLinkOption',
+			onSelect: () => {
+				fallback();
+			},
```
We could simplify this line by doing:
```suggestion
			onSelect: fallback,
```

jd-alexander replied:
Agreed! a2dc77b

fluiddot reviewed (commented):
Awesome work @jd-alexander 🥇 !

I added some suggestions and here I list some items that would be nice to review:

- On Android, the `Retry` and `Convert to link` buttons are not left-aligned as in the original designs:
<img width="350" alt="Screenshot 2021-09-10 at 13 07 36" src="https://user-images.githubusercontent.com/14905380/132844745-2a82804b-6c0e-4f41-a993-87fe87d34e68.png">

- In the `Unable to embed media` state, there's no way to edit the URL. I think we should enable the edit URL button in the toolbar (Pencil icon) or add a new option for this purpose in the picker, otherwise, the only option to modify it would be by removing and adding an Embed block again.
https://github.com/WordPress/gutenberg/blob/7b31b96760b490dfbda6cb505534671c0d7b58d7/packages/block-library/src/embed/edit.native.js#L243

fluiddot commented:
> Looks like we need to review the error handling on the native side because we should receive at least the status code when the request fails. I also noticed that the api-fetch-setup.js file, which processes the requests initiated from the RN side, expects the status code for handling the error so this might be a regression 🤔 .

I realized that at least on iOS, we're already receiving the status code so omit my comment. I also noticed that we already have a timeout mechanism implemented here:
https://github.com/WordPress/gutenberg/blob/2ea08a512cf0b8c1553314ceeb6545e8c1354841/packages/react-native-editor/src/api-fetch-setup.js#L42-L44

However, it might take too long to trigger the error because by default it retries 20 times:
https://github.com/WordPress/gutenberg/blob/2ea08a512cf0b8c1553314ceeb6545e8c1354841/packages/react-native-editor/src/api-fetch-setup.js#L20

I tested this with Charles Proxy and dropping the connection for `https://public-api.wordpress.com/oembed/1.0/sites/*` URLs, it took around 4 minutes until the error was triggered and set the state to `Unable to embed media`.

jd-alexander commented:
> However, it might take too long to trigger the error because by default it retries 20 times:

Indeed! I had seen that retry logic. Thanks for bringing it up!

> I tested this with Charles Proxy and dropping the connection for https://public-api.wordpress.com/oembed/1.0/sites/* URLs, it took around 4 minutes until the error was triggered and set the state to Unable to embed media.

That is a long wait time. I am trying to think of a workaround that we could do. One idea would be to create a `DISABLED_TIMEOUT_ENDPOINTS` that may be counterproductive since retries are appropriate as a fallback mechanism when requests fail. We could also lessen the retries count for these URLs using the same matching behavior. I am not fond of these solutions but I am just putting them out there while I continue to think about a solution.

jd-alexander commented:
By the way, just to double-check here because since we are receiving the iOS error code would we then be able to utilize that if the error code is similar on Android to compute a result? I will take another look at the Android side of things.

> At this point, we should focus on solving the missing error status code and figure out a way to detect the availability of the internet connection to cover the network-related errors. But I don't think this should block the PR as we could address it when tackling the Offline support for Embeds blocks issue, wdyt?

This answers the question :) Once we have the error status codes from both platforms we can unify the behavior of how we handle the result of the `getEmbedPreview` call. 

> But I don't think this should block the PR as we could address it when tackling the Offline support for Embeds blocks issue, wdyt? 

Indeed. We don't have to block this PR as the focus of it is not to modify the underlying API requests behavior :) I will make the respective changes for another review.

jd-alexander committed:
Merge branch 'trunk' into rnmobile/embed-block-error-bottomsheet

jd-alexander committed:
Performed left align of Picker options.

jd-alexander committed:
Changed the bottom sheet title to reflect the design spec

jd-alexander committed:
Simplified function assignment in picker options.

jd-alexander committed:
Removed onPickerPresent function and simplified presentPicker access.

jd-alexander committed:
Added the Edit Button EmbedControls

jd-alexander committed:
If cannotEmbed then the accessibility hint reflections the embed options

jd-alexander committed:
removed space before the colon in changelog entry.

jd-alexander commented:
> Awesome work @jd-alexander 🥇 !

Thanks for the review 🙇🏾 

> On Android, the Retry and Convert to link buttons are not left-aligned as in the original designs:

1. True! I aligned the options here [bacd89e](https://github.com/WordPress/gutenberg/pull/34604/commits/bacd89e7c99e06f682047fa9e01e935c39447582)

2. I also changed the title of the bottom sheet as well to match the designs. [6184d10](https://github.com/WordPress/gutenberg/pull/34604/commits/6184d10f21985ceb495b27109552ddbc4c46eab5) 

<kbd><img src="https://user-images.githubusercontent.com/1509205/132960379-cdff0532-2db2-4164-bf98-9d2cabffe3bf.png" width="320"></kbd>

> In the Unable to embed media state, there's no way to edit the URL. I think we should enable the edit URL button in the toolbar (Pencil icon) or add a new option for this purpose in the picker, otherwise, the only option to modify it would be by removing and adding an Embed block again.

Since the current implementation utilizing the edit URL button in the toolbar I did that implementation for consistency. We could do a rain check on this though to see if it would be better to put "Edit link URL" within the Picker as well for discoverability and usability purposes. 

I added the control here [1dc892b](https://github.com/WordPress/gutenberg/pull/34604/commits/1dc892be819cddee9a0f586c1818a86ea592d143). It is only shown when `cannotEmbed` is true. 

<kbd><img src="https://user-images.githubusercontent.com/1509205/132960329-272d562e-09ca-414d-85d7-2ee4f3dcd6a5.png" width="320"></kbd>

fluiddot commented:
Thanks for addressing the issues I mentioned in the review ❤️ !

I've tested them and works perfectly:
- On Android, the Retry and Convert to link buttons are not left-aligned as in the original designs ✅ 
- Change the title of the bottom sheet to match the designs ✅ 
- Show the edit URL button in the toolbar when `cannotEmbed` is `true` ✅ 

> Since the current implementation utilizing the edit URL button in the toolbar I did that implementation for consistency. We could do a rain check on this though to see if it would be better to put "Edit link URL" within the Picker as well for discoverability and usability purposes.

I agree, for consistency, I rather display the edit URL button in the toolbar. Regarding including it in the picker, we could ask for design input in a private channel.

fluiddot reviewed (approved):
LGTM 🎊 !

_Tested on Simulator iPhone 12 Pro Max (iOS 14.5) and Samsung Galaxy S20 FE 5G (Android 10)._

Regarding one of the initial comments related to the picker title:
> The design contains a Picker that has a description. The current Picker implementation does not support this hence I opted to show only the error description which is a generic hardcoded value that the web Embed component allows displays.

With the changes introduced in [this commit](https://github.com/WordPress/gutenberg/pull/34604/commits/6184d10f21985ceb495b27109552ddbc4c46eab5), I noticed that now the picker's title is `Embed options` instead of `Sorry, this content could not be embedded.`. I'm ok leaving it this way but I'm wondering if we should introduce the error message eventually to the picker component to match the designs, probably we could open an issue as a follow-up, wdyt?

<img src=https://user-images.githubusercontent.com/1509205/132262888-ded29ad6-843f-4cdc-b69b-09ca6d428f88.png width=350>

> Also, I noticed that the existing Error UI in the Placeholder is Red instead of Dark Grey. Should this be changed?

Good point, I see that in the original designs the color used is Dark Grey which makes more sense for the text `Embed options`, but I agree that in case we introduce the error message `Sorry, this content could not be embedded.`, it would make more sense to use a red one.

fluiddot commented:
As a side note, while testing I bumped into the following issue:
- [Embed block: URL is not updated when edited after setting a bad URL of a provider](https://github.com/WordPress/gutenberg/issues/34792)

jd-alexander commented:
Thanks for the review @fluiddot 😄 

> I agree, for consistency, I rather display the edit URL button in the toolbar. Regarding including it in the picker, we could ask for design input in a private channel.

Agreed!

> With the changes introduced in this commit, I noticed that now the picker's title is Embed options instead of Sorry, this content could not be embedded.. I'm ok leaving it this way but I'm wondering if we should introduce the error message eventually to the picker component to match the designs, probably we could open an issue as a follow-up, wdyt?

Sounds like a good approach to me. I created an issue for the Picker description prop https://github.com/wordpress-mobile/gutenberg-mobile/issues/3962 and an issue to track the implementation of this error message in the Error BottomSheet so that this PR isn't blocked https://github.com/wordpress-mobile/gutenberg-mobile/issues/3963

> Good point, I see that in the original designs the color used is Dark Grey which makes more sense for the text Embed options, but I agree that in case we introduce the error message Sorry, this content could not be embedded., it would make more sense to use a red one.

Sounds good to me.

> As a side note, while testing I bumped into the following issue:
> Embed block: URL is not updated when edited after setting a bad URL of a provider

Thanks for filing 🙇🏾

jd-alexander committed:
Merge branch 'trunk' into rnmobile/embed-block-error-bottomsheet

jd-alexander committed:
Merge branch 'trunk' into rnmobile/embed-block-error-bottomsheet

jd-alexander committed:
Merge branch 'trunk' into rnmobile/embed-block-error-bottomsheet
